### PR TITLE
Fix Issues 19728, 20093 - AliasSeq in UDA errors

### PR DESCRIPTION
**Pull request by BorisCarvajal**
Not sure if it's the best approach but it works.

### CODE REVIEW

**Comment by UplinkCoder** [comment]
This looks dangerous.

**Comment by BorisCarvajal** [comment]
I felt the same yet nothing was broken.
This tuple interaction seems very rare, maybe with your new `type functions` more cases will pop up.

File: src/dmd/expression.d [hunk 1]
```diff
@@ -260,6 +260,14 @@ extern (C++) void expandTuples(Expressions* exps)
...
                     continue;
                 }
+                else // Expand a TypeTuple
+                {
+                    exps.remove(i);
+                    auto texps = new Expressions();
+                    foreach (a; *tt.arguments)
+                        texps.push(new TypeExp(e.loc, a.type));
+                    exps.insert(i, texps);
+                }
```
Comment (Geod24) [inline src/dmd/expression.d:270]: https://github.com/dlang/dmd/blob/f4f2db858f0eb6938bb09b88abda4e1cb1337d6b/src/dmd/expression.d#L231-L238

**Comment by Geod24** [comment]
Could you explain what was the problem, and how you solved it ? It's pretty hard to review otherwise.

**Comment by BorisCarvajal** [comment]
The tuples in the uda were not expanding if they were all only types.

Still each element is enclosed by a tuple though, as it's the current behavior.

File: test/compilable/test19728.d [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+@Seq!(A,Seq!(B,A),B) struct Foo5 {}
+@Seq!(A,Seq!(B,A),B) @Seq!(A,B,A,B) struct Foo6 {}
+
+pragma(msg, __traits(getAttributes, Foo1));
+pragma(msg, __traits(getAttributes, Foo2));
+pragma(msg, __traits(getAttributes, Foo3));
+pragma(msg, __traits(getAttributes, Foo4));
+pragma(msg, __traits(getAttributes, Foo5));
+pragma(msg, __traits(getAttributes, Foo6));
+static assert(is(typeof(__traits(getAttributes, Foo1))));
```
Comment (atilaneves) [inline test/compilable/test19728.d:32]: There should be tests to verify what the UDAs actually are.

Comment (BorisCarvajal) [reply test/compilable/test19728.d]: Is better now?

**Comment by BorisCarvajal** [comment]
Don't merge this yet.

File: src/dmd/expression.d [hunk 3]
```diff
@@ -260,6 +260,14 @@ extern (C++) void expandTuples(Expressions* exps)
                     i--;
                     continue;
                 }
+                else // Expand a TypeTuple
+                {
+                    exps.remove(i);
+                    auto texps = new Expressions();
+                    foreach (a; *tt.arguments)
```
Comment (UplinkCoder) [inline src/dmd/expression.d:267]: Use setLength please and assign to the index.

**Review by UplinkCoder — CHANGES_REQUESTED** [review]

**Comment by UplinkCoder** [comment]
I've had some time to think about this.
This might be sound.

File: src/dmd/expression.d (hunk 3 above)
Comment (UplinkCoder) [reply src/dmd/expression.d:267]: What you are doing right now is Equivalent to `~=` in a loop.
Which should be avoided if the number of elements inserted per iteration is known and the loop is bounds are known you can preallocate.

**Comment by BorisCarvajal** [comment]
I'm fixing the problem in another place now, but I still think this code path should be there (expanding a TypeTuple within a TupleExp). The case with an empty TypeTuple is already there.

File: src/dmd/expression.d (hunk 3 above)
Comment (BorisCarvajal) [reply src/dmd/expression.d:267]: Right, should preallocate.

File: test/compilable/test19728.d (hunk 2 above)
Comment (atilaneves) [reply test/compilable/test19728.d]: Yes, thanks.

**Comment by thewilsonator** [comment]
So is this good to go?

**Comment by BorisCarvajal** [comment]
No yet, I need to push some changes.

**Comment by thewilsonator** [comment]
Is this good now? If not ping me when it is.

File: src/dmd/expression.d (hunk 3 above)
Comment (BorisCarvajal) [reply src/dmd/expression.d]: Done.

**Pull request merged**